### PR TITLE
key schedule reorg

### PR DESCRIPTION
**Pull request by franziskuskiefer**
In preparation for the key schedule test vectors I re-organized the
key schedule and moved the logic into the actual key schedule module
instead of the group.

There are no changes to functionality PR, only moving code around.

This PR also removes the test-vector feature as it is no longer needed.

### CODE REVIEW

File: src/schedule/mod.rs [hunk 1]
```diff
@@ -43,26 +43,96 @@
 //! welcome_secret) is represented by its own struct to ensure that the keys are
 //! not confused with one-another and/or that the schedule is not derived
 //! out-of-order.
+//!
+//! ## The real key schedules
+//! The spec isn't great here. The key schedule is actually
```
Comment (raphaelrobert) [inline src/schedule/mod.rs:48]: These kind of comments are not very helpful for others. 

From what I gather, the difference between the two key schedules is only the beginning, namely:

```
                  init_secret_[n-1]
                         |                       
                         V                         
    commit_secret -> KDF.Extract                         
                         |
                         V                         
                   DeriveSecret(., "joiner")
                         |
                         V
```

I think it would be much clearer to explain the discrepancy between the spec and the implementation.

**Review by raphaelrobert — CHANGES_REQUESTED** [review]
I see some functional changes in this PR, in the sense that previously the type safety guaranteed a certain order of execution which has now been replaced with an internal state. In general this is more error prone, but I think it's tolerable in this case. 
I also understand the need for the refactoring, because previously it was impossible to extract data for test vectors.
Mostly nits, with some questions about the actual key schedule part.

File: src/group/mls_group/new_from_welcome.rs [hunk 2]
```diff
@@ -32,13 +31,27 @@ impl MlsGroup {
...
+        let group_secrets = GroupSecrets::decode_detached(&group_secrets_bytes)?;
+        let joiner_secret = group_secrets.joiner_secret;
+        // TODO #141: Implement PSK
+        let mut key_schedule = KeySchedule::init(ciphersuite, joiner_secret, None);
+        let (welcome_key, welcome_nonce) = key_schedule
+            .welcome()?
+            .derive_welcome_key_nonce(ciphersuite);
+
+        let group_info_bytes =
+            match welcome_key.aead_open(welcome.encrypted_group_info(), &[], &welcome_nonce) {
```
Comment (kkohbrok) [inline src/group/mls_group/new_from_welcome.rs:49]: I think `ok_or(...)` would be a bit nicer here.

File: src/schedule/mod.rs (hunk 1 above)
Comment (kkohbrok) [reply src/schedule/mod.rs:48]: I'm curious why you think the extract-expand of the commit and init secrets into the joiner secret is not actually part of the key schedule. Can you elaborate on that?

**Review by kkohbrok — COMMENTED** [review]
This should work. I'd still prefer a type safe variant without the need for runtime errors, but this might be the more rusty way. If found [this](https://hoverbear.org/blog/rust-state-machine-pattern/) post, which contains an approach that which might be a compromise that uses type safety, but does a better job of hiding the logic from the outside.

File: src/group/mls_group/create_commit.rs [hunk 3]
```diff
@@ -215,3 +202,67 @@ impl MlsGroup {
...
+        invited_members: Vec<(NodeIndex, AddProposal)>,
+        provisional_tree: &RatchetTree,
+    ) -> Result<Vec<Self>, CodecError> {
+        // Get a Vector containing the node indices of the direct path to the
+        // root from our own leaf.
+        let dirpath = treemath::leaf_direct_path(
+            provisional_tree.own_node_index().into(),
+            provisional_tree.leaf_count(),
+        )
+        .expect("create_commit_internal: TreeMath error when computing direct path.");
```
[the remark this answers is not in the document]
Comment (franziskuskiefer) [reply src/group/mls_group/create_commit.rs:227]: Sure I'll fix this. There's more here we might want to do. But this code is only moved and I don't want to touch too much of it in this PR.

File: src/group/mls_group/new_from_welcome.rs (hunk 2 above)
Comment (franziskuskiefer) [reply src/group/mls_group/new_from_welcome.rs:49]: Sure, will do (`map_err` it is actually). But same comment as above. This code is only moved around and I don't want to touch it too much in this PR.

File: src/schedule/mod.rs (hunk 1 above)
Comment (franziskuskiefer) [reply src/schedule/mod.rs:48]: Well there are both flows you everyone can just compare to see the differences. Also, I'd expect this to change in the spec before its done. And then we should update the comment here.
I updated the description here to explain a little more.

File: src/schedule/mod.rs [hunk 4]
```diff
@@ -92,95 +162,126 @@ impl JoinerSecret {
...
+        }
+
+        Ok(InitSecret::new(
+            self.ciphersuite,
+            self.epoch_secret.as_ref().unwrap(),
+        ))
+    }
+
+    /// Derive the epoch secrets.
+    pub(crate) fn epoch_secrets(&mut self) -> Result<EpochSecrets, KeyScheduleError> {
```
[the remark this answers is not in the document]
Comment (franziskuskiefer) [reply src/schedule/mod.rs:261]: There is no order to these secrets. There are only two different functions because we don't always need the `init_secret`. If we want to get the additional guarantee that we can't derive secrets multiple times, we can do so with adding more complexity here. I added the additional state.
But note that first deriving the `EpochSecrets` and then the `InitSecrets` is the right order if the state machine should terminate. We don't always derive the `init_secret`, which would leave the state machine in an open state.
The other option to solve this is integrating the `init_secret` into the `EpochSecrets` struct and only deriving it when requested.

**Review by franziskuskiefer — COMMENTED** [review]
Thanks for the comments. I'll elaborate a little on the need for this and the issues with the key schedule in the spec.

First, the refactoring does not in any change the type safety guarantees of the secrets. This all works the same way as before. The state machine only has additional checks.
We could certainly put more effort into making the state machine nicer. But it's so small I don't think it's worth the effort.

I started the refactoring in order to implement test vectors. But this is needed in any case to slowly move OpenMLS towards somewhat more maintainable code.

The key schedule as described in the spec isn't really one key schedule. The `joiner_secret` is an intermediate value *and* an output value. This must never be the case within a key schedule. The actual key schedule is therefore only the second half starting with the `joiner_secret`, which indeed is what happens when starting a group from a welcome message.

File: src/schedule/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (kkohbrok) [reply src/schedule/mod.rs:299]: An alternative could be modelling the key schedule as a state machine with a loop, representing the chain that it is in the spec. That would mean it permanently lives in the group state, but that could be a viable option.

File: src/schedule/mod.rs [hunk 5]
```diff
@@ -225,15 +366,15 @@ impl WelcomeSecret {
...
     secret: Secret,
 }
 
 impl EpochSecret {
     /// Derive an `EpochSecret` from a `JoinerSecret`
-    pub(crate) fn new(
+    fn new(
         ciphersuite: &Ciphersuite,
-        intermediate_secret: IntermediateSecret,
+        intermediate_secret: &IntermediateSecret,
```
Comment (kkohbrok) [inline src/schedule/mod.rs:377]: Can't we consume the intermediate secret here? I believe that should be possible if we use `take()` in `add_context`.

File: src/group/mls_group/new_from_welcome.rs (hunk 2 above)
Comment (kkohbrok) [reply src/group/mls_group/new_from_welcome.rs]: I thought you use `map_err` if you want to use the existing error to compute the new one. `ok_or` just ignores the existing error by default and just outputs the one you specify.

**Review by kkohbrok — CHANGES_REQUESTED** [review]
> The key schedule as described in the spec isn't really one key schedule. The `joiner_secret` is an intermediate value _and_ an output value. This must never be the case within a key schedule. The actual key schedule is therefore only the second half starting with the `joiner_secret`, which indeed is what happens when starting a group from a welcome message.

I see what you mean. It would certainly be nice not to have to "output" the `joiner_secret`, but I think some such step is unfortunately required to allow new members to join the key schedule chain. Do you have an idea to make the structure nicer? You could also view it a bit differently, such that the joiner secret of the next epoch is the "end" of the key schedule as opposed to `init_secret_[n]`.

File: src/schedule/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (raphaelrobert) [reply src/schedule/mod.rs:299]: From all secrets, the init secret is the most sensitive one, because if it leaks it allows an attacker to break into a group permanently. That's why I am concerned about it being derived more than once. 
You have a point about the fact, that we don't always derive the init secret though. I personally don't care to much about the state machine terminating, being able to derive the init secret only once is more important to me. If we want the state machine to always terminate I guess we would need two different functions:
1. Derive the epoch secrets and terminate
2. Derive the epoch secrets, then the init secret and then terminate.

File: src/group/mls_group/create_commit.rs [hunk 6]
```diff
@@ -221,10 +224,10 @@ impl PlaintextSecret {
         // Get a Vector containing the node indices of the direct path to the
         // root from our own leaf.
         let dirpath = treemath::leaf_direct_path(
-            provisional_tree.own_node_index().into(),
+            provisional_tree.own_node_index(),
             provisional_tree.leaf_count(),
         )
-        .expect("create_commit_internal: TreeMath error when computing direct path.");
+        .unwrap();
```
Comment (raphaelrobert) [inline src/group/mls_group/create_commit.rs:230]: This is still missing the comment on why the `unwrap()` is safe. I agree it's not part of the PR per-se, but let's finish this correctly.

File: src/tree/mod.rs [hunk 7]
```diff
@@ -460,9 +462,9 @@ impl RatchetTree {
         key_package_bundle.set_key_package(key_package.clone());
 
         (
-            self.private_tree.commit_secret(),
+            // self.private_tree.commit_secret(),
```
Comment (raphaelrobert) [inline src/tree/mod.rs:465]: comments

**Review by raphaelrobert — CHANGES_REQUESTED** [review]
Thanks for the extensive explanation, the key schedule looks good now.

Just a few nits otherwise.

File: src/schedule/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (franziskuskiefer) [reply src/schedule/mod.rs:299]: Ok, I don't actually think this is super important. Deleting the key material is more important. But the cleanest option then is to do it all at the same time (the `init_secret` optionally). I'll do that.

**Review by kkohbrok — APPROVED** [review]
Looks good to me. At some point, we will want to keep some of the epoch secrets around, while other (e.g. the `init_secret`) should be deleted for FS. Do we have an issue for that already?

**Comment by franziskuskiefer** [comment]
> Looks good to me. At some point, we will want to keep some of the epoch secrets around, while other (e.g. the `init_secret`) should be deleted for FS. Do we have an issue for that already?

Yeah we certainly need to look at some FS aspects at some point.... There are some issues for some of it already. But I don't think there's one for the epoch secrets.

**Review by raphaelrobert — APPROVED** [review]
Thanks, LGTM!

I agree about the FS concerns, I think we have #335 for that.
